File: lidar_proc/src/lidar_proc.py
#!/usr/bin/env python
import roslib; roslib.load_manifest('lidar_proc')
import rospy
import math
from std_msgs.msg import Float64
from sensor_msgs.msg import LaserScan, PointCloud
from geometry_msgs.msg import Point32
from filters.msg import EKFData

pub = rospy.Publisher('point_cloud_raw', PointCloud)
pub_acc = rospy.Publisher('point_cloud_raw_acc', PointCloud)
angle = Float64()
ekf = EKFData()
MAX_DISTANCE = 4
p_acc = PointCloud()
p_acc.points = []
p.header.frame_id = "map"
def angleCallback(data):
    global angle
    angle = data

def ekfCallback(data):
    global ekf
    ekf = data

def lidarCallback(data):
    global angle
    global ekf
    global p_acc
    p = PointCloud()
    p.points = []
    p.header = data.header
    p.header.frame_id = "map"
    p_acc.header.stamp = data.header.stamp
    p_acc.header.seq = data.header.seq
    current_angle = data.angle_min
    pitch = angle.data + ekf.pitch
    roll = ekf.roll
    for i in data.ranges:
        if i < MAX_DISTANCE:
            point = Point32()
            yaw = current_angle+ekf.yaw
            point.x = i * math.cos(yaw) * math.cos(pitch)
            point.y = i * math.sin(yaw) * math.cos(pitch)
            current_angle += data.angle_increment
            point.z = i * math.sin(pitch)
            p.points += [point]
    p_acc.points += p.points
    pub.publish(p)
    pub_acc.publish(p_acc)

def lidar():
    rospy.init_node('point_cloud_raw', anonymous=False)
    rospy.Subscriber("scan", LaserScan, lidarCallback)
    rospy.Subscriber("hokuyo_angle", Float64, angleCallback)
    rospy.Subscriber("ekf_data", EKFData, ekfCallback)
    rospy.spin()

if __name__ == '__main__':
    lidar()

File: lidar_proc/src/lidar_proc.cpp
#include <ros/ros.h>
#include <iostream>
#include <stdio.h>
#include <vector>
#include <math.h>
#include <std_msgs/String.h>
#include <sensor_msgs/LaserScan.h>
#include <sensor_msgs/PointCloud.h>
#include <geometry_msgs/Point32.h>
#include <laser_geometry/laser_geometry.h>

using namespace std;

class LidarToPointCloud {
    public:
        LidarToPointCloud();
        void scanCallback(const sensor_msgs::LaserScan::ConstPtr& scan);
    private:
        ros::NodeHandle node_;
        laser_geometry::LaserProjection projector_;

        ros::Publisher point_cloud_publisher_;
        ros::Subscriber scan_sub_;
};

LidarToPointCloud::LidarToPointCloud() {
    scan_sub_ = node_.subscribe<sensor_msgs::LaserScan> ("/scan", 100, &LidarToPointCloud::scanCallback, this);

    point_cloud_publisher_ = node_.advertise<sensor_msgs::PointCloud>("/cloud", 100, false);
}

void LidarToPointCloud::scanCallback (const sensor_msgs::LaserScan::ConstPtr& scan) {
    sensor_msgs::PointCloud cloud;
    projector_.projectLaser(*scan, cloud);

    point_cloud_publisher_.publish(cloud);
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "lidar_processing");

    LidarToPointCloud filter;

    ros::spin();

    return 0;
}
/*
void lidarinfo (const sensor_msgs::LaserScanPtr& msg) {

    double current_angle = msg->angle_min;
    sensor_msgs::PointCloud p();
    int size = msg->ranges.size();
    vector<geometry_msgs::Point32> ptsvec();
    for (int i = 0; i < size; i++)
    {
        geometry_msgs::Point32 newPoint();
        newPoint.x = msg->ranges[msg->ranges[i]] * cos(current_angle);
        newPoint.y = msg->ranges[msg->ranges[i]] * sin(current_angle);
        current_angle += msg->angle_increment;
        newPoint.z = 0;
        ptsvec.push_back(newPoint);
    }
    p.points = &ptsvec;
    point_pub.publish(p);
    */
